### ISSUE TITLE
Union mount attempts to write to a read-only upstream

### ISSUE THREAD

**Issue by robotoshi** [rclone/rclone#4929]
#### What is the problem you are having with rclone?

According to the documentation for the Union backend:

> - All **action** policies will filter out remotes which are tagged as **read-only**.
> - All **create** policies will filter out remotes which are tagged **read-only** or **no-create**.

But this does not appear to always be obeyed.
If I mount a union of a read-only upstream and a writeable local upstream, then modify a file that is only present on the read-only one, rclone keeps on trying to upload that change to the read-only upstream and then failing because it's read-only!

What I would *really* like is for it to do a copy-on-write and just create that modified file on my local upstream, but I think I saw that as a really old feature request. Most importantly thoughis for it to quit trying to upload to a remote that it knows is read-only.

Flags `--retries`, `--retries-sleep`, and `--low-level-retries` all appear to have no effect on th number or duration of the retries.

I've tried rc API calls `vfs/forget`, `vfs/refresh` as well as sending a SIGHUP, but they do not actually clear the cache enough to completely foget about the modified file. 
(Side note: `cache/expire` returns a 404)

#### What is your rclone version (output from `rclone version`)

```
rclone v1.53.3
- os/arch: linux/amd64
- go version: go1.15.5
```
Installed using the most recent .deb

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Ubuntu 20.04.1 LTS "Focal" (64-bit)

####  Which cloud storage system are you using? (e.g. Google Drive)

Local+Local
and
Local+HTTP

#### The command you were trying to run

`rclone mount -vv merge: /home/user/my-mount --vfs-cache-mode=writes`

#### The rclone config contents with secrets removed.

```
  [my-local]
  type = local
  
  [my-remote]
  type = http
  url = http://localhost:8080
  
  [merge]
  type = union
  upstreams = my-local:/home/user/dir my-remote:/:ro
  action_policy = all
  create_policy = all
  search_policy = all
```

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
2021/01/10 22:17:33 DEBUG : read-only-file.txt: vfs cache: starting upload
2021/01/10 22:17:33 ERROR : read-only-file.txt: Failed to copy: permission denied
2021/01/10 22:17:33 ERROR : read-only-file.txt: vfs cache: failed to upload try #1, will retry in 10s: vfs cache: failed to transfer file from cache to remote: permission denied
...
```

#### Workaround

The only workaround I've found to make it forget about the files and restore the originals from the remote is to manually delete the file from the VFS cache (`rm ~/.cache/vfs/path/to/file`), at which point it will attempt to "upload" it again and "succeed"

```
2021/01/13 18:40:19 DEBUG : vfs cache RemoveNotInUse (maxAge=3600000000000, emptyOnly=false): item path/to/file not removed, freed 0 bytes
2021/01/13 18:40:19 INFO  : vfs cache: cleaned: objects 1 (was 1) in use 1, to upload 1, uploading 0, total size 104 (was 104)
2021/01/13 18:40:19 DEBUG : path/to/file: vfs cache: starting upload
2021/01/13 18:40:19 DEBUG : path/to/file: vfs cache: writeback object to VFS layer
2021/01/13 18:40:19 DEBUG : : Added virtual directory entry vAdd: "H.h"
2021/01/13 18:40:19 INFO  : path/to/file: vfs cache: upload succeeded try #2
2021/01/13 18:41:19 DEBUG : vfs cache RemoveNotInUse (maxAge=3600000000000, emptyOnly=false): item path/to/file not removed, freed 0 bytes
2021/01/13 18:41:19 INFO  : vfs cache: cleaned: objects 1 (was 1) in use 0, to upload 0, uploading 0, total size 104 (was 104)
```
And that's basically fine.  Seems like it's still claiming to keep the file in its cache, but the working directory is restored and it doesn't keep retrying.
Note how it still reports the file as "not removed".
It appears the only way to make it *truly* "forget" about the file is to restart rclone.

**Comment by robotoshi** [comment]
Just tested on the latest beta
```
rclone v1.54.0-beta.5062.db2c38b21
- os/arch: linux/amd64
- go version: go1.15.6
```
and got all the same results

**Comment by ncw** [comment]
This can be replicated without the VFS like this

```
$ rclone config show uniontest
--------------------
[uniontest]
type = union
upstreams = /tmp/rw /tmp/ro:ro
action_policy = all
create_policy = all
search_policy = all
--------------------
$ mkdir /tmp/rw
$ mkdir /tmp/ro
$ echo hello > /tmp/ro/file.txt
$ rclone lsf uniontest:
file.txt
$ rclone cat uniontest:
hello
$ echo newfile > file.txt
$ rclone -vv copy file.txt uniontest: --retries 1 --low-level-retries 1
2021/01/14 11:14:45 DEBUG : rclone: Version "v1.54.0-beta.5061" starting with parameters ["rclone" "-vv" "copy" "file.txt" "uniontest:" "--retries" "1" "--low-level-retries" "1"]
2021/01/14 11:14:45 DEBUG : Creating backend with remote "file.txt"
2021/01/14 11:14:45 DEBUG : Using config file from "/home/ncw/.rclone.conf"
2021/01/14 11:14:45 DEBUG : fs cache: adding new entry for parent of "file.txt", "/home/ncw"
2021/01/14 11:14:45 DEBUG : Creating backend with remote "uniontest:"
2021/01/14 11:14:45 DEBUG : Creating backend with remote "/tmp/ro"
2021/01/14 11:14:45 DEBUG : Creating backend with remote "/tmp/rw"
2021/01/14 11:14:45 DEBUG : union root '': actionPolicy = *policy.All, createPolicy = *policy.All, searchPolicy = *policy.All
2021/01/14 11:14:45 DEBUG : file.txt: Sizes differ (src 8 vs dst 6)
2021/01/14 11:14:45 ERROR : file.txt: Failed to copy: permission denied
2021/01/14 11:14:45 ERROR : Attempt 1/1 failed with 1 errors and: permission denied
2021/01/14 11:14:45 INFO  : 
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 1 (retrying may help)
Elapsed time:         0.0s

2021/01/14 11:14:45 DEBUG : 2 go routines active
2021/01/14 11:14:45 Failed to copy: permission denied
```

What is happening is that rclone is attempting to Update the Object rather than creating one from scratch.

Creating one from scratch would do the whole copy on write thing.

I've attempted to implement that here - can you give it a go?

[v1.54.0-beta.5062.236aa5a70.fix-4929-union](https://beta.rclone.org/branch/fix-4929-union/v1.54.0-beta.5062.236aa5a70.fix-4929-union/) on branch [fix-4929-union](https://github.com/rclone/rclone/tree/fix-4929-union) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
@robotoshi - did you have a chance to have a go with the beta above?
